src/main/java/teammates/storage/sqlapi/UserDb.java

```diff
@@ -0,0 +1,138 @@
+package teammates.storage.sqlapi;
+
+import java.time.Instant;
+
+import org.hibernate.Session;
+
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.HibernateUtil;
+import teammates.storage.sqlentity.Instructor;
+import teammates.storage.sqlentity.Student;
+import teammates.storage.sqlentity.User;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+
+/**
+ * Handles CRUD operations for users.
+ *
+ * @see User
+ */
+public final class UserDb extends EntitiesDb<User> {
+    private static final UserDb instance = new UserDb();
+
+    private UserDb() {
+        // prevent initialization
+    }
+
+    public static UserDb inst() {
+        return instance;
+    }
+
+    /**
+     * Creates a user.
+     */
+    public <T extends User> T createUser(T user)
+            throws InvalidParametersException, EntityAlreadyExistsException {
+        assert user != null;
+
+        if (!user.isValid()) {
+            throw new InvalidParametersException(user.getInvalidityInfo());
+        }
+
+        if (getInstructor(user.getId()) == null) {
+            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
+        }
+
+        if (getStudent(user.getId()) == null) {
+            throw new EntityAlreadyExistsException(String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, user.toString()));
+        }
```

Comment:
If it's a student we want to persist, then getInstructor will be null and an exception will be thrown.